dali/operators/ssd/box_encoder.cc

```diff
@@ -80,74 +80,88 @@ vector<std::pair<unsigned, unsigned>> BoxEncoder<CPUBackend>::MatchBoxesWithAnch
   return matches;
 }
 
-vector<BoundingBox> BoxEncoder<CPUBackend>::ReadBoxesFromInput(
+vector<Box<2, float>> BoxEncoder<CPUBackend>::ReadBoxesFromInput(
   const float *in_boxes, unsigned num_boxes) const {
-  vector<BoundingBox> boxes;
-  boxes.reserve(num_boxes);
-  auto in_box_data = in_boxes;
-
+  vector<Box<2, float>> boxes;
+  boxes.resize(num_boxes);
+  const float *in_box_data = in_boxes;
   for (unsigned idx = 0; idx < num_boxes; ++idx) {
-    boxes.push_back(BoundingBox::FromLtrb(in_box_data, BoundingBox::NoBounds()));
-    in_box_data += BoundingBox::kSize;
+    auto &box = boxes[idx];
+    box.lo[0] = in_box_data[0];
+    box.lo[1] = in_box_data[1];
+    box.hi[0] = in_box_data[2];
+    box.hi[1] = in_box_data[3];
+    in_box_data += kBboxSize;
   }
 
   return boxes;
 }
 
-void BoxEncoder<CPUBackend>::WriteBoxToOutput(const std::array<float, BoundingBox::kSize> &box,
-                                              float *out_box_data) const {
-  for (unsigned idx = 0; idx < BoundingBox::kSize; ++idx)
-    out_box_data[idx] = box[idx];
+template <int ndim>
+void WriteBoxToOutput(float *out_box_data, const vec<ndim, float> &center,
```

Comment:
Previously the (center, extent) pair was represented as [Bounding]Box. That would allow to utilize the WriteBoxes function if I'm not mistaken. Why this split?

Reply:
I'd say that was an abuse of box structure - the box's members have well defined semantics and we shouldn't just use "box" as a container because it is extremely error prone and a nightmare to read.